validator/client/validator.go

```diff
@@ -369,6 +370,97 @@ func (v *validator) SlotDeadline(slot types.Slot) time.Time {
 	return time.Unix(int64(v.genesisTime), 0 /*ns*/).Add(secs * time.Second)
 }
 
+// CheckDoppelGanger checks if the current actively provided keys have
+// any duplicates active in the network.
+func (v *validator) CheckDoppelGanger(ctx context.Context) error {
+	if !featureconfig.Get().EnableDoppelGanger {
+		return nil
+	}
+	pubkeys, err := v.keyManager.FetchValidatingPublicKeys(ctx)
+	if err != nil {
+		return err
+	}
+	log.WithField("keys", len(pubkeys)).Info("Running doppelganger check")
+	// Exit early if no validating pub keys are found.
+	if len(pubkeys) == 0 {
+		return nil
+	}
+	req := &ethpb.DoppelGangerRequest{ValidatorRequests: []*ethpb.DoppelGangerRequest_ValidatorRequest{}}
+	for _, pkey := range pubkeys {
+		attRec, err := v.db.AttestationHistoryForPubKey(ctx, pkey)
+		if err != nil {
+			return err
+		}
+		if len(attRec) == 0 {
+			// If no history exists we simply send in a zero
+			// value for the request epoch and root.
+			req.ValidatorRequests = append(req.ValidatorRequests,
+				&ethpb.DoppelGangerRequest_ValidatorRequest{
+					PublicKey:  pkey[:],
+					Epoch:      0,
+					SignedRoot: make([]byte, 32),
+				})
+			continue
+		}
+		r := retrieveLatestRecord(attRec)
+		if pkey != r.PubKey {
+			return errors.New("attestation record mismatched public key")
+		}
```

Comment:
Added this as a sanity check. It may not be necessary but it was some assurance that we received the correct record.